spring-integration-stomp/src/main/java/org/springframework/integration/stomp/support/StompHeaderMapper.java

```diff
@@ -0,0 +1,235 @@
+					name = entry1.getKey();
+					if (shouldMapHeader(name, this.outboundHeaderNames)) {
+						String value = entry1.getValue().get(0);
+						if (StringUtils.hasText(value)) {
+							setStompHeader(target, name, value);
+						}
+					}
+				}
+			}
+		}
+	}
+
+	private void setStompHeader(StompHeaders target, String name, Object value) {
+		if (StompHeaders.CONTENT_LENGTH.equals(name)) {
+			if (value instanceof Number) {
+				target.setContentLength(((Number) value).longValue());
+			}
+			else if (value instanceof String) {
+				target.setContentLength(Long.parseLong((String) value));
+			}
+			else {
+				Class<?> clazz = (value != null) ? value.getClass() : null;
+				throw new IllegalArgumentException(
+						"Expected Number or String value for 'content-length' header value, but received: " + clazz);
+			}
+		}
+		else if (StompHeaders.CONTENT_TYPE.equals(name) || MessageHeaders.CONTENT_TYPE.equals(name)) {
+			MimeType contentType = target.getContentType();
+			if (contentType == null || StompHeaders.CONTENT_TYPE.equals(name)) {
+				if (value instanceof MediaType) {
+					target.setContentType((MediaType) value);
+				}
+				else if (value instanceof String) {
+					target.setContentType(MediaType.parseMediaType((String) value));
+				}
+				else {
+					Class<?> clazz = (value != null) ? value.getClass() : null;
+					throw new IllegalArgumentException(
+							"Expected MediaType or String value for 'content-type' header value, but received: " + clazz);
+				}
+			}
+		}
+		else if (StompHeaders.DESTINATION.equals(name)) {
+			if (value instanceof String) {
+				target.setDestination((String) value);
+			}
+			else {
+				Class<?> clazz = (value != null) ? value.getClass() : null;
+				throw new IllegalArgumentException(
+						"Expected String value for 'destination' header value, but received: " + clazz);
+			}
+		}
+		else if (StompHeaders.RECEIPT.equals(name)) {
+			if (value instanceof String) {
+				target.setReceipt((String) value);
+			}
+			else {
+				Class<?> clazz = (value != null) ? value.getClass() : null;
+				throw new IllegalArgumentException(
+						"Expected String value for 'receipt' header value, but received: " + clazz);
+			}
+		}
+		else {
+			if (value instanceof String) {
+				target.set(name, (String) value);
+			}
+			else {
+				Class<?> clazz = (value != null) ? value.getClass() : null;
+				throw new IllegalArgumentException(
+						"Expected String value for any generic STOMP header value, but received: " + clazz);
+			}
+		}
+	}
+
+	@Override
+	public Map<String, Object> toHeaders(StompHeaders source) {
+		Map<String, Object> target = new HashMap<String, Object>();
+		for (String name : source.keySet()) {
+			if (shouldMapHeader(name, this.inboundHeaderNames)) {
+				if (StompHeaders.CONTENT_TYPE.equals(name)) {
+					target.put(MessageHeaders.CONTENT_TYPE, source.getContentType());
+				}
+				else {
+					target.put(name, source.getFirst(name));
```

Comment:
Why only the first value?

Also, shouldn't we be prefixing the headers `stomp_` ??